cpp/src/Slice/MetadataValidation.h

```diff
@@ -0,0 +1,101 @@
+    /// should return a string describing the error. This string will be supplied to the user as part of a diagnostic.
+    using ValidationFunc = std::function<std::optional<std::string>(const MetadataPtr&, const SyntaxTreeBasePtr&)>;
+
+    struct MetadataInfo
+    {
+        /// A list of types that this metadata can validly be applied to.
+        ///
+        /// If this list is empty we don't perform any automatic checking of whether this metadata is validly applied.
+        /// Usually, this is because determining validity isn't as straightforward as matching against a list,
+        /// and requires a more complex approach, which is achieved through providing an `extraValidation` function.
+        std::list<std::reference_wrapper<const std::type_info>> validOn;
+
+        /// Specifies how many, and what kinds of arguments, this metadata accepts.
+        MetadataArgumentKind acceptedArguments;
+
+        /// This field stores the specific values that can be provided as arguments to this metadata.
+        /// If this field is unset, then we perform no validation of the arguments (i.e. arguments can have any value).
+        std::optional<StringList> validArgumentValues = std::nullopt;
+
+        /// Specifies in what contexts the metadata can appear (i.e. can it apply to defintions, references, both?)
+        MetadataApplicationContext acceptedContexts = MetadataApplicationContext::Definitions;
+
+        /// Indicates whether it's valid and meaningful for this metadata to appear multiple times on the same thing.
+        bool mustBeUnique = true;
+
+        /// A function used to run additional validation for this metadata.
+        /// If this field is set, it will always be run.
+        /// @see ValidationFunc
+        ValidationFunc extraValidation = nullptr;
+    };
+
+    /// Returns a message of the form "'<directive>' metadata cannot be applied to <type>'" (more or less).
+    /// This message is reported to the user when metadata is placed on something for which it is inapplicable.
+    std::string misappliedMetadataMessage(const MetadataPtr& metadata, const SyntaxTreeBasePtr& p);
+
+    /// Validates all the metadata in the provided Unit against a map of known metadata.
+    /// @param p The unit who's metadata should be validated.
+    /// @param language Which language's metadata should be validated. Any metadata that starts with a different
+    ///                 language prefix than this one will be ignored. Note that metadata without any language prefix
+    ///                 (i.e. parser metadata) is always checked.
+    /// @param knownMetadata A map containing the directives that should be validated, and information describing
+    ///                      the various constraints and conditions that should be upheld for it and its arguments.
+    void validateMetadata(const UnitPtr& p, std::string language, std::map<std::string, MetadataInfo> knownMetadata);
```

Comment:
const& since the implementation only reads these parameters.

Reply:
I changed the `language` to a `string_view` (since it's preferred over `const string&`)

But the `knownMetadata` map was left as-is, since the implementation does actually modify it, and I'd like to avoid making a working copy. Especially since the caller has no use for the map; it exists solely to be consumed by this function.

And for completeness, the function as a whole shouldn't be `const` because it has side-effects (invalid metadata is filtered out, so the code-gen is assured it will only encounter valid metadata).